crates/core/src/error/summary.rs

```diff
@@ -0,0 +1,415 @@
+//! An informative summary system for aggregating and condensing data collected
+//! from runtime checks, including warnings, issues, and operational metrics.
+//!
+//! This system should provide end-users with a clear, concise summary of command
+//! execution results without conflicting with existing error-handling standards.
+//! In scenarios where execution cannot proceed due to a critical error, a
+//! `RusticError` will be raised instead, and no summary will be provided.
+//!
+//! # Separation of Concerns
+//!
+//! Critical runtime errors that prevent further execution are handled through the
+//! existing `RusticError` system. The `Summary` will only collect information for
+//! non-fatal events.
+//!
+//! # Compatibility with Existing Error Handling
+//!
+//! Summaries must coexist with error propagation rules. They will not replace
+//! the core behavior of error propagation but act as a complementary mechanism
+//! for presenting non-fatal feedback.
+//!
+//! # User-Friendly Reporting
+//!
+//! Summaries should aggregate detailed runtime information—such as warnings,
+//! issues, and metrics — in a clear and condensed format for the end-user.
+//!
+//! # Aggregation & Condensation
+//!
+//! Similar or repeated errors should be aggregated to avoid redundant information,
+//! presenting users with a high-level overview.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    fmt::{self, Display},
+    time::Instant,
+};
+
+use ecow::EcoString;
+
+pub type IssueIdentifier = EcoString;
+
+pub type Issues = BTreeMap<IssueScope, BTreeMap<IssueIdentifier, CondensedIssue>>;
+pub type Metrics = BTreeMap<EcoString, EcoString>;
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Default,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    derive_more::Display,
+    serde::Serialize,
+)]
+pub enum IssueScope {
+    #[default]
+    Internal,
+    Unknown,
+    UserInput,
+}
+
+#[derive(Debug, Clone, Default, serde::Serialize)]
+pub struct CondensedIssue {
+    /// High-level description of the problem
+    message: EcoString,
+
+    /// Number of occurrences
+    count: usize,
+
+    /// Optional diagnostic information, e.g. an error message
+    root_cause: Option<EcoString>,
+}
+
+#[derive(
+    Debug, Clone, Copy, Default, Hash, PartialEq, Eq, derive_more::Display, serde::Serialize,
+)]
+pub enum DisplayOptionKind {
+    #[default]
+    Issues,
+    Timing,
+    Metrics,
+    All,
+}
+
```

Comment:
These could be bitflags or enumflags, but I'm unsure if that wouldn't be overengineering?